### DISCUSSION TITLE
No persistent WiFi on Yi Outdoor (4CCN YHS.3017), any reboot needs new WiFi QR code scan.

### DISCUSSION THREAD

**Discussion by MichlW** [alienatedsec/yi-hack-v5#315] [Q&A] [answered]
Short version: "Welcome to yi home camera" after any reboot.

Hi everyone,

I am from germany and got 4 pieces of chinese version of Yi Outdoor cameras (4CCN YHS.3017). Because I am familiar with yi-hack projects (v3, v4, v5) and using it on another Yi outdoor camera for several years now, I tried to hack my 4 new CN cams.

The cameras are compatible with yi-hack-v5 and I was able to flash them using a 32GB FAT32 card as expected. After that it was possible to integrate the cams to my WiFi by using the international app version of Yi Home or an online QR code generator. I can reach the hack on port 8080 and configure the cams.

My problem is: After any reboot of the cams, they are not included in my WiFi anymore. Anytime they come with that **"Welcome to yi camera"** audio and I should scan the QR code again. I suppose that this is because the pairing can not be finished completely with my app version.

So i decided to try to configure the WiFi credentials in "configure_wifi.cfg" within yi-hack, but this also does not work. Then I tried to use proxychains-ng to make the camera believe it is in China, but also this does not help.

So anytime I tried the next thing lets me believe, that it is essential to completely pair the cams with the chinese version of the Yi Home app - is this correct? I read some threads that mentioned to use Yi cams with these hack without initially using the app. Is it essential to pair the cam with the app and is it essential to do this before or after applying the hack?

May someone assist me in completing the list of steps to install yi-hack on chinese outdoor cams?

1) Copy yi-hack-v5 folder and both home and root files from 0.3.8 to a FAT32 formatted sd cards.
2) Insert the card in the camera and wait until the hack is completely flashed.
3) Initially pair the cams with my WiFi using the QR code from app or website.
(Can I then ignore the pairing timeout from that CN-INT-mismatch?)
3a) check and/or re-configure WiFi credentials through my browser.
4) Update the cam to 0.4.1b pre-release from browser ui button with the mandatory reboot.
5) Re-pair the cams to my WiFi using the QR code from app or website.
6) Configure and test proxychains-ng through the browser.

What is missing that the WiFi connection not persists a reboot? What I am doing wrong?

Disclaimer: I can not easily try chinese app because I have no access to an android phone and we only use iPhones and iPads. For me, it would be the best solution if I don't need the app at all.

Thanks in advance!

**Comment by alienatedsec** [comment]
@MichlW There is a script in the `/home/app/scripts/wificonnect.sh` - please can you share its content?

**Reply by MichlW** [reply]
@alienatedsec After a fresh flash/install of 0.3.8 and also after upgrading to 0.4.1b the script content is:
```
ifconfig wlan0 up
sleep 1
/home/base/tools/wpa_supplicant -c/tmp/wpa_supplicant.conf -g/var/run/wpa_supplicant-global -iwlan0 -B;
```

and the content of /tmp/wpa_supplicant.conf is:
```
ctrl_interface=/var/run/wpa_supplicant
ap_scan=1
network={
ssid="<correct but deleted by me>"
scan_ssid=1
proto=WPA RSN
key_mgmt=WPA-PSK
pairwise=CCMP TKIP
group=CCMP TKIP
psk="<correct but deleted by me>"
}
```

**Reply by MichlW** [reply]
I tested alot and I now understand, that I first need a working proxy to bypass region restrictions and AFTER that I should try to pair with the chinese version of the iOS app. So I figured it out to configure proxychains-ng and got this proof:
```
/home/yi-hack-v5 # proxychains4 wget -O- -q http://ipinfo.io
[proxychains] config file found: /tmp/sd/yi-hack-v5/etc/proxychains.conf
[proxychains] preloading /tmp/sd/yi-hack-v5/lib/libproxychains4.so
[proxychains] DLL init: proxychains-ng 4.16
[proxychains] Random chain  ...  183.56.203.183:1080  ...  34.117.59.81:80  ...  OK
{
  "ip": "183.56.203.183",
  "city": "Shenzhen",
  "region": "Guangdong",
  "country": "CN",
  "loc": "22.5455,114.0683",
  "org": "AS58466 CHINANET Guangdong province network",
  "timezone": "Asia/Shanghai",
  "readme": "https://ipinfo.io/missingauth"
}
```
This is a stable result and then I also figured it out to create an chinese Apple-ID and install the chinese version of the Yi Home app to an old iPhone SE (1st gen).

But even with that working proxy configuration I am not able to pair the cam with the app. The voice says "Pairing timed out. Please reset the camera".

What have I missed?

**Reply by alienatedsec** [reply]
Not sure yet but there is a proxychain web interface that could be useful.

**Reply by MichlW** [reply]
The results there are also correct:

![image](https://github.com/alienatedsec/yi-hack-v5/assets/49558935/365de0f9-f84d-40b3-be48-ec5581be4353)

but pairing with the app fails over and over.
Even after trying these "optimizations": [yi-hack-MStar/issues/187](https://github.com/roleoroleo/yi-hack-MStar/issues/187)

**Reply by alienatedsec** [reply]
> @alienatedsec After a fresh flash/install of 0.3.8 and also after upgrading to 0.4.1b the script content is:
> 
> ```
> 
> ifconfig wlan0 up
> 
> sleep 1
> 
> /home/base/tools/wpa_supplicant -c/tmp/wpa_supplicant.conf -g/var/run/wpa_supplicant-global -iwlan0 -B;
> 
> ```
> 
> 
> 

Add the above onto system.sh on your SD and reboot

> and the content of /tmp/wpa_supplicant.conf is:
> 
> ```
> 
> ctrl_interface=/var/run/wpa_supplicant
> 
> ap_scan=1
> 
> network={
> 
> ssid="<correct but deleted by me>"
> 
> scan_ssid=1
> 
> proto=WPA RSN
> 
> key_mgmt=WPA-PSK
> 
> pairwise=CCMP TKIP
> 
> group=CCMP TKIP
> 
> psk="<correct but deleted by me>"
> 
> }
> 
> ```


If still doesn't work, create the same config file on your SD. Modify the command from the system.sh to point at your configuration file.

**Reply by MichlW** [reply]
Thank you, but both suggestions did not help. There is no different behavior after doing this. Still not able to pair the cam with the app.

I now also tried to connect the cam and the iPhone with the chinese app to a "clean" mobile phone hotspot instead of my local WiFi. Also this does not work. So DNS adblocker or firewall in OpenWRT are not the cause of my pairing problems.

My last idea is to try an android phone with chinese apk installation - The problem: I do not have such a phone ;-)
Therefore I thought about using an Android emulator for PC, but I have no idea how to set correct WiFi connection in an emulated Android.

**Reply by alienatedsec** [reply]
@MichlW Apologies, I can't help any further as I have no Chinese models in hand to test.

**Reply by MichlW** [reply]
@alienatedsec I can donate one of my four 4CCN models to you and this project. Are you from germany?

**Reply by alienatedsec** [reply]
> @alienatedsec I can donate one of my four 4CCN models to you and this project. Are you from Germany?

I am based in the UK @MichlW

**Reply by MichlW** [reply]
@alienatedsec I figured it out to pair the camera with the app by giving yi-hack-v4 a try. But it was only successful after changing some lines in `/home/app/cloudAPI` as suggested in [yi-hack-v4/issues/37](https://github.com/TheCrypt0/yi-hack-v4/issues/37#issuecomment-485868655) from
```
if [[ $(get_config PROXYCHAINSNG) == "yes" ]] ; then
       $YI_HACK_V4_PREFIX/bin/proxychains4 ./cloudAPI_real "$@"
else
       ./cloudAPI_real "$@"
fi
```
to
```
if [ "$4" == "https://api.xiaoyi.com/v4/ipc/check_did" ]; then
        echo '{"allow":true,"code":"20000"}'
elif [[ $(get_config PROXYCHAINSNG) == "yes" ]] ; then
        $YI_HACK_V4_PREFIX/bin/proxychains4 ./cloudAPI_real "$@"
else
        ./cloudAPI_real "$@"
fi
```
Then I thought, that I can reach the goal also on yi-hack-v5, but it does not work as in v4. I made changes in the files `/home/app/cloudAPI`, `/home/yi-hack-v5/script/cloudAPI` and in `/tmp/sd/yi-hack-v5/script/cloudAPI`, because I wasn't sure which one is the right one.

I changed them from
```
if [[ $(get_config DISABLE_CLOUD) == "yes" ]] ; then
        cd /home/app
        ./cloudAPI_fake "$@"
else
    if [[ $(get_config PROXYCHAINSNG) == "yes" ]] ; then
        $YI_HACK_V5_PREFIX/bin/proxychains4 ./home/app/cloudAPI_real "$@"
    else
        cd /home/app
        ./cloudAPI_real "$@"
    fi
fi
```
to
```
if [[ $(get_config DISABLE_CLOUD) == "yes" ]] ; then
        cd /home/app
        ./cloudAPI_fake "$@"
else
    if [ "$4" == "https://api.xiaoyi.com/v4/ipc/check_did" ]; then
        echo '{"allow":true,"code":"20000"}'
    elif [[ $(get_config PROXYCHAINSNG) == "yes" ]] ; then
        $YI_HACK_V5_PREFIX/bin/proxychains4 ./home/app/cloudAPI_real "$@"
    else
        cd /home/app
        ./cloudAPI_real "$@"
    fi
fi
```
Do you have an idea why this does not work as in v4? Thanks in advance.

**Reply by MichlW** [reply]
And is there a way to pair the camera in v4 and then flash it to v5 and it remains paired with the app?

**Reply by alienatedsec** [reply]
@MichlW I will ask before doing anything - did you turn the proxychain setting on?

If you cannot connect to the camera, you will need to amend `system.conf` manually

**Reply by MichlW** [reply, marked as answer]
> And is there a way to pair the camera in v4 and then flash it to v5 and it remains paired with the app?

That worked! The camera now persists the WiFi connection between reboots following these steps:

1. Flash the camera with yi-hack-v4 version 0.2.0
2. Connect the cam to WiFi with the chinese version of the Yi Home app - without completing the pairing process!
3. Add working proxies to proxychains.conf and enable proxychains-ng. (Test with `proxychains4 wget -O- -q http://ipinfo.io`)
4. Add the following code to `/home/app/cloudAPI`:
```
if [ "$4" == "https://api.xiaoyi.com/v4/ipc/check_did" ]; then
        echo '{"allow":true,"code":"20000"}'
elif [...
```
5. Reboot the cam
6. Connect the cam to WiFi with the chinese version of the Yi Home app and now complete the pairing process. This should work now after that cloudAPI changes.
7. Insert a sd card with yi-hack-v5 baseline version 0.3.8 and reflash the cam.
(I decided to directly disable cloud features and enabling proxychains-ng in `system.conf` prior to flashing yi-hack-v5. I also added these working proxies to `proxychains.conf` prior to flashing)
9. Update the cam to latest version (i.e. pre-release 0.4.1c)

I will do this with three other CN cams in the next days and update these instructions if necessary.

@alienatedsec Thanks for all your support!
Unfortunately, I do not currently understand why pairing in yi-hack-v5 does not work after making the cloudAPI changes there as well.

**Reply by MichlW** [reply]
> @MichlW I will ask before doing anything - did you turn the proxychain setting on?
> 
> If you cannot connect to the camera, you will need to amend `system.conf` manually

Yes, of course!

**Reply by alienatedsec** [reply]
@MichlW 
I found the problem ;)
The fix is coming in the next release :)

**Reply by MichlW** [reply]
@alienatedsec These are wonderful news. I can test prior to release with my china outdoor cams, if you want.

**Reply by alienatedsec** [reply]
@MichlW 

```
if [[ $(get_config DISABLE_CLOUD) == "yes" ]] ; then
        cd /home/app
        ./cloudAPI_fake "$@"
else
    if [ "$4" == "https://api.xiaoyi.com/v4/ipc/check_did" ]; then
        echo '{"allow":true,"code":"20000"}'
    elif [[ $(get_config PROXYCHAINSNG) == "yes" ]] ; then
        cd /home/app
        $YI_HACK_V5_PREFIX/bin/proxychains4 ./cloudAPI_real "$@"
    else
        cd /home/app
        ./cloudAPI_real "$@"
    fi
fi
```

**Reply by alienatedsec** [reply]
@MichlW Did you try `0.4.1c` (according to your comments, yes) or `0.4.1d` yet? Check this #318

**Reply by MichlW** [reply]
@alienatedsec I tried 0.4.1d and it looks that pairing is possible. I now have other strange behavior: Camera says "pairing successful", then it says "invalid qr code" and then it says "pairing timed out" despite the fact that the camera is visible in the yi home app.

I retried pairing with yi-hack-v4 as described in [comment 315](https://github.com/alienatedsec/yi-hack-v5/discussions/315#discussioncomment-7639760) above where the app pairing requires only 10sec and is much faster than the 2min with v5 0.4.1d and does not show the strange behavior described above.

But at the moment I have no clue why this is the case.
